src/basicfuns.jl

```diff
@@ -261,4 +261,4 @@ $(SIGNATURES)
 Return the [`softmax transformation`](https://en.wikipedia.org/wiki/Softmax_function)
 applied to `x`.
 """
-softmax(x::AbstractArray{<:Real}) = softmax!(similar(x, Float64), x)
+softmax(x::AbstractArray{R<:Real}) = softmax!(similar(x, R), x)
```

Comment:
I suggest
```suggestion
softmax(x::AbstractArray{<:Real}) = softmax!(similar(x, float(eltype(x))), x)
```
This avoids the type parameter (and the missing `where` clause) and should work with integervalued arrays as well.

Reply:
Can you also add a test to `test/basicfuns.jl` that checks that it works for integer-valued arrays? And maybe Float32 as expected, i.e., returns an array of element type Float32?